File: Structural/Composite/Composite.kt
// Implementation

interface Shape {
    fun draw(fillColor: String)
}

class Square: Shape {
    override fun draw(fillColor: String) {
        println("Drawing a Square with color" + fillColor)
    }
}

class Circle: Shape {
    override fun draw(fillColor: String) {
        println("Drawing a Circle with color" + fillColor)
    }
}

class WhiteBoard(var shapes: ArrayList<Shape>): Shape {
    override fun draw(fillColor: String) {
        for (shape in shapes) {
            shape.draw(fillColor)
        }
    }
}

// Usage

val shapes = arrayListOf<Shape>(Circle(), Square())
var whiteboard = WhiteBoard(shapes)
whiteboard.draw("Red")

File: Structural/Composite/composite.swift
// Implementation

protocol Shape {
    func draw(fillColor: String)
}
/*: 
Leafs
*/ 
final class Square : Shape {
    func draw(fillColor: String) {
        print("Drawing a Square with color \(fillColor)")
    }
}

final class Circle : Shape {
    func draw(fillColor: String) {
        print("Drawing a circle with color \(fillColor)")
    }
}

/*:
Composite
*/
final class Whiteboard : Shape {
    lazy var shapes = [Shape]()
    
    init(_ shapes:Shape...) {
        self.shapes = shapes
    }
    
    func draw(fillColor: String) {
        for shape in self.shapes {
            shape.draw(fillColor: fillColor)
        }
    }
}

// Usage:

var whiteboard = Whiteboard(Circle(), Square())
whiteboard.draw("Red")
